### ISSUE TITLE
Office for Web dialog API not initializing Office.onReady, preventing call to context.ui.messageParent

### ISSUE THREAD

**Issue by justmars** [OfficeDev/office-js#3673]
## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: Office on the web
* Host [Excel, Word, PowerPoint, etc.]: Word
* Office version number: 16.0.16912.41013 (build)
* Operating System: Mac
* Browser (if using Office on the web): Edge, Chrome, Firefox
 
## Expected behavior

1. Dialog box is launched from taskpane with a host page `localhost:3000/taskpane` containing the following script:
    ```js
    Office.onReady().then(() => {
        Office.context.ui.displayDialogAsync(
        connectDialogUrl, // https://localhost:3000/accounts/login
        { height: 70, width: 50 },
        function (result) {
          console.log(JSON.stringify(result));
          if (result.status === Office.AsyncResultStatus.Succeeded) {
            if (dialog) {
              dialog.close()
            };
            dialog = result.value;
            dialog.addEventHandler(Office.EventType.DialogMessageReceived, function(arg) => {
              dialog.close();
              if (arg.message === "true") {
                window.location.replace('https://localhost:3000/target'); // redirect taskpane
              }
            } );
          }
        })})
    ```
3. Dialog box url  `https://localhost:3000/accounts/login` contains `host_Info` in the querystring
4. The html is the dialog box url does not contain a reference to `<script src="https://appsforoffice.microsoft.com/lib/1/hosted/office.js"></script>`... since including this results in the dialog box immediately closing. See what I mean by this through [a scriptlab snippet](https://github.com/OfficeDev/script-lab/issues/1017)
5. Through the dialog box, user logs in either via username/password or oAuth2
6. If successful, the terminal page contains a script to reference office js and invocation of such script:
    ```html
    <html>
      <head>
        <script src="https://appsforoffice.microsoft.com/lib/1/hosted/office.js"></script>
      </head>
      <h1>hello world</h1>
    </html>
    <script>
      Office.onReady().then(() => Office.context.ui.messageParent("true"))
    </script>
    ```
7. The above script should open the dialog then closes the same, navigating the host page to a target url: `https://localhost:3000/target`

## Current behavior

Instead of the expected last step, the dialog box remains open and contains the target url. 

Note that in office for Mac, the above `Expected behavior` works.

## Steps to reproduce

1. Create the same script above
2. Sideload the manifest file in office for web
3. Launch the dialog
4. Allow user to login via such dialog
5. The url in the dialog should terminate in a page that contains the officejs script
6. The terminal page should "talk" back to the host page, so that the host page can close the dialog

# Provide additional details

See detailed discovery in #3586 

## Useful logs

- [x] Console errors
- [ ] Screenshots
- [ ] Test file (if only happens on a particular file)

The dialog box doesn't appear to read Office.onReady() after a successful login. The url of the terminal login no longer contains the querystring including the `host_Info`. The result in different browsers using Word, Office for Web:

### Edge

`Uncaught Sys.ArgumentNullException: Sys.ArgumentNullException: Value cannot be null. Parameter name: conversationId`

### Chrome

`Warning: Office.js is loaded outside of Office client`
`Uncaught TypeError: Office.context.ui is undefined`

### Firefox

`Warning: Office.js is loaded outside of Office client`
`Uncaught TypeError: Office.context.ui is undefined`

Thank you for taking the time to report an issue. Our triage team will respond to you in less than 72 hours. Normally, response time is <10 hours Monday through Friday. We do not triage on weekends.

**Comment by justmars** [comment]
Apparently, my localhost was serving the dialog url with the response header `Cross-Origin-Opener-Policy: same-origin-allow-popups`. Once i switched this off, the dialog url was able to properly load the OfficeJS reference in the head.

**Comment by jdglaser** [comment]
I am experiencing this exact same issue now when trying to handle an Oauth2.0 redirect flow with an Excel add-in. @justmars, can you elaborate on how you were able to solve this?

**Comment by jgreywolf** [comment]
@jdglaser Did you ever resolve your issue?

**Comment by fds-megcaballero** [comment]
@jdglaser @jgreywolf , I was experiencing the same issue but for me the HTML I was opening in the dialog had `Cross-Origin-Opener-Policy: same-origin` header in the response. After completely removing that header, I was able to show the page in dialog without issues.

**Comment by jim22k** [comment]
@jdglaser I found a workaround for this problem by avoiding any redirects in the dialog. Instead, you pop open a second window, do the OAuth flow there, and communicate back to the first dialog using BroadcastChannel. I created a working example here: https://github.com/jim22k/OfficeOnlineDialog/tree/main

The only downside to this approach is that most browsers will block the 2nd popup window, so users will need to enable popups for your add-in domain. Or you will need to have a button on the first dialog that they can click to open the 2nd window and do the auth flow.

Hope this helps.
